### PR TITLE
chore: use CarStream in `cli car concat`

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

As follow-up of https://github.com/ChainSafe/forest/pull/3325#discussion_r1284252504

Changes introduced in this pull request:

- Use CarStream in `cli car concat`

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/cli/subcommands/car_cmd.rs [hunk 1]
```diff
@@ -61,29 +62,28 @@ impl CarCommands {
...
-}
-
-fn dedup_block_stream(stream: impl Stream<Item = BlockPair>) -> impl Stream<Item = BlockPair> {
+fn dedup_block_stream(
+    stream: impl Stream<Item = std::io::Result<Block>>,
+) -> impl Stream<Item = BlockPair> {
     let mut seen = CidHashSet::default();
-    stream.filter(move |(cid, _data)| futures::future::ready(seen.insert(*cid)))
+    stream.filter_map(move |result| {
+        // FIXME: Remove `.expect` once `CarHeader` `write` API accepts `TryStream`
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:79]: Let's either use the ForestCAR.zst encoder or write a sink for plain CAR data.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: src/cli/subcommands/car_cmd.rs [hunk 2]
```diff
@@ -30,71 +27,67 @@ impl CarCommands {
...
+    futures::stream::iter(car_streams).flatten()
 }
 
-fn dedup_block_stream(stream: impl Stream<Item = BlockPair>) -> impl Stream<Item = BlockPair> {
+fn dedup_block_stream(
+    stream: impl Stream<Item = std::io::Result<Block>>,
+) -> impl Stream<Item = std::io::Result<Block>> {
     let mut seen = CidHashSet::default();
-    stream.filter(move |(cid, _data)| futures::future::ready(seen.insert(*cid)))
+    stream.try_filter_map(move |block| {
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:71]:
```suggestion
    stream.try_filter(move |block| {
```

File: src/cli/subcommands/car_cmd.rs [hunk 3]
```diff
@@ -122,12 +115,16 @@ mod tests {
             car
         }
 
-        fn into_stream(self) -> impl Stream<Item = BlockPair> {
-            futures::stream::iter(self.0.into_iter().map(|b| (b.cid, b.data)))
+        fn into_stream(self) -> impl Stream<Item = std::io::Result<super::Block>> {
+            futures::stream::iter(
+                self.0
+                    .into_iter()
+                    .map(|Block { cid, data }| Ok::<_, std::io::Error>(super::Block { cid, data })),
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:122]: We map block into block?

File: tests/car_tests.rs [hunk 4]
```diff
@@ -123,10 +124,13 @@ fn new_block(rng: &mut SmallRng) -> (Cid, Vec<u8>) {
...
 
 async fn validate_car(path: impl AsRef<Path>) -> Result<()> {
-    let mut reader = CarReader::new(tokio::fs::File::open(path).await?.compat()).await?;
-    assert!(reader.validate);
+    let mut reader = CarStream::new(tokio::io::BufReader::new(
+        tokio::fs::File::open(path).await?,
+    ))
+    .await?;
+    assert!(!reader.header.roots.is_empty());
     let mut count = 0;
```
Comment (lemmih) [inline tests/car_tests.rs:132]:
```suggestion
    let count = reader.count();
```

File: src/cli/subcommands/car_cmd.rs (hunk 3 above)
Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs:122]: From `fvm_ipld_car::Block` to `crate::utils::db::car_stream::Block`

Comment (lemmih) [reply src/cli/subcommands/car_cmd.rs:122]: Why use `fvm_ipld_car::Block`?

File: tests/car_tests.rs (hunk 4 above)
Comment (hanabi1224) [reply tests/car_tests.rs]: Hmm, this does not work, `count()` takes the ownership of self. I need to call `while reader.try_next()` to validate all blocks

File: src/cli/subcommands/car_cmd.rs (hunk 3 above)
Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs:122]: Need to use `CarHeader` API to convert it to car bytes. I don't find equivalent API in our own `CarHeader` tho

```rust
async fn into_car_bytes(self) -> Vec<u8> {
            // Dummy root
            let writer = CarHeader::from(vec![self.0[0].cid]);
            let mut car = vec![];
            let mut stream = pin!(futures::stream::iter(self.0).map(|b| (b.cid, b.data)));
            writer
                .write_stream_async(&mut car, &mut stream)
                .await
                .unwrap();
            car
        }
```

Comment (lemmih) [reply src/cli/subcommands/car_cmd.rs:116]: Then let's implement our own CAR sink. Building on top of `CarHeader` is no use.

File: tests/car_tests.rs (hunk 4 above)
Comment (lemmih) [reply tests/car_tests.rs:132]: So you don't actually care about the count. Looks like you just want to collect the stream with `try_collect()`. If you want the count for debugging then you can use `inspect_ok()`.

File: src/cli/subcommands/car_cmd.rs (hunk 3 above)
Comment (lemmih) [reply src/cli/subcommands/car_cmd.rs:116]: Or simply generate ForestCAR.zst bytes.

Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs]: Nvm, actually it's OK to use own `Block` by updating the roundtrip test, fixed.

File: src/utils/db/car_index/car_index_builder.rs [hunk 5]
```diff
@@ -141,4 +141,8 @@ impl CarIndexBuilder {
     pub fn len(&self) -> u64 {
         self.table.len() as u64
     }
+
+    pub fn is_empty(&self) -> bool {
+        self.table.is_empty()
+    }
```
Comment (lemmih) [inline src/utils/db/car_index/car_index_builder.rs:147]: Where is this function used?

Comment (hanabi1224) [reply src/utils/db/car_index/car_index_builder.rs:147]: This is to fix a clippy warning after `mod util` has been made public to be accessible in tests/*

```
warning: struct `CarIndexBuilder` has a public `len` method, but no `is_empty` method
   --> src/utils/db/car_index/car_index_builder.rs:141:5
    |
141 |     pub fn len(&self) -> u64 {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_without_is_empty
    = note: `#[warn(clippy::len_without_is_empty)]` on by default
```

File: src/cli/subcommands/car_cmd.rs [hunk 6]
```diff
@@ -2,25 +2,22 @@
...
-
-type BlockPair = (Cid, Vec<u8>);
+use crate::utils::db::car_stream::{Block, CarStream};
 
 #[derive(Debug, Subcommand)]
 pub enum CarCommands {
     Concat {
-        /// A list of `.car` file paths
+        /// A list of CAR file paths. A CAR file can be a plain CAR, a zstd compressed CAR
+        /// or a `ForestCAR.zst`
```
Comment (elmattic) [inline src/cli/subcommands/car_cmd.rs:18]: What's a `ForestCAR.zst`? I suppose it's a `.forest.car.zst` file. Maybe using extensions supported here would be clearer.

**Review by elmattic — APPROVED** [review]

File: src/cli/subcommands/car_cmd.rs (hunk 6 above)
Comment (lemmih) [reply src/cli/subcommands/car_cmd.rs:18]: We can use the file extension here.

Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs]: Fixed

File: src/utils/db/car_index/car_index_builder.rs (hunk 5 above)
Comment (lemmih) [reply src/utils/db/car_index/car_index_builder.rs:147]: Let's add an exemption. This definition of `is_empty` is not meaningful.

Comment (lemmih) [reply src/utils/db/car_index/car_index_builder.rs:147]: Can also rename `len` to `capacity`.

Comment (hanabi1224) [reply src/utils/db/car_index/car_index_builder.rs]: Renamed to `capacity`

File: src/cli/subcommands/car_cmd.rs [hunk 7]
```diff
@@ -30,71 +27,61 @@ impl CarCommands {
...
+    use std::pin::pin;
+
     use super::*;
     use ahash::HashSet;
     use cid::multihash;
     use cid::multihash::MultihashDigest;
     use cid::Cid;
     use futures::executor::{block_on, block_on_stream};
-    use fvm_ipld_car::Block;
+    use fvm_ipld_car::CarHeader;
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:84]: Let's get rid of `CarHeader` completely.

File: src/cli/subcommands/car_cmd.rs [hunk 8]
```diff
@@ -155,9 +142,9 @@ mod tests {
     fn blocks_roundtrip(blocks: Blocks) -> anyhow::Result<()> {
         block_on(async move {
             let car = blocks.into_car_bytes().await;
-            let mut reader = CarReader::new(car.as_slice()).await?;
+            let mut reader = CarStream::new(std::io::Cursor::new(car.clone())).await?;
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:145]:
```suggestion
            let mut reader = CarStream::new(std::io::Cursor::new(&car)).await?;
```

File: src/cli/subcommands/car_cmd.rs [hunk 9]
```diff
@@ -155,9 +142,9 @@ mod tests {
...
             let mut blocks2 = vec![];
-            while let Some(b) = reader.next_block().await? {
+            while let Some(b) = reader.try_next().await? {
                 blocks2.push(b);
             }
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:149]:
```suggestion
            let blocks2 = Blocks(reader.try_collect().await?);
```

File: src/cli/subcommands/car_cmd.rs [hunk 10]
```diff
@@ -204,15 +191,15 @@ mod tests {
...
-                CarReader::new(car_a.as_slice()).await?,
-                CarReader::new(car_b.as_slice()).await?,
+            let car_a = std::io::Cursor::new(a.into_car_bytes().await);
+            let car_b = std::io::Cursor::new(b.into_car_bytes().await);
+            let mut deduped = pin!(dedup_block_stream(merge_car_streams(vec![
+                CarStream::new(car_a).await?,
+                CarStream::new(car_b).await?,
             ])));
 
             let mut cid_union2 = HashSet::default();
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:201]:
```suggestion
            let cid_union2: HashSet<Cid> = deduped.map_ok(|block| block.cid).try_collect().await?;
```

File: src/cli/subcommands/car_cmd.rs [hunk 11]
```diff
@@ -2,25 +2,22 @@
...
 
 #[derive(Debug, Subcommand)]
 pub enum CarCommands {
     Concat {
-        /// A list of `.car` file paths
+        /// A list of CAR file paths. A CAR file can be a plain CAR, a zstd compressed CAR
+        /// or a `.forest.car.zst` file
         car_files: Vec<PathBuf>,
-        /// The output `.car` file path
+        /// The output `ForestCAR.zst`(`.forest.car.zst`) file path
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:20]:
```suggestion
        /// The output `.forest.car.zst` file path
```

File: src/cli/subcommands/car_cmd.rs (hunk 7 above)
Comment (lemmih) [reply src/cli/subcommands/car_cmd.rs:84]: We should either switch to ForestCAR.zst or write our own sink.

File: tests/car_tests.rs [hunk 12]
```diff
@@ -10,9 +10,11 @@ use cid::{
     multihash::{self, MultihashDigest},
     Cid,
 };
-use futures::StreamExt;
-use fvm_ipld_car::{CarHeader, CarReader};
+use forest_filecoin::utils::db::car_stream::CarStream;
+use futures::{StreamExt, TryStreamExt};
+use fvm_ipld_car::CarHeader;
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply tests/car_tests.rs]: Done

File: tests/car_tests.rs [hunk 13]
```diff
@@ -99,36 +102,45 @@ async fn forest_cli_car_concat_same_file_3_times() -> Result<()> {
...
     let rng = SmallRng::seed_from_u64(0xdeadbeef);
-    let (cid, _data) = new_block(&mut rng.clone());
-    let header = CarHeader::from(vec![cid]);
-
-    let mut block_stream = Box::pin(
-        futures::stream::unfold(rng, |mut rng| async { Some((new_block(&mut rng), rng)) })
-            .take(size),
+    let root_block = new_block(&mut rng.clone());
+    let roots = vec![root_block.cid];
+    let block_stream = futures::stream::iter(vec![Ok(root_block)]).chain(
```
Comment (lemmih) [inline tests/car_tests.rs:107]: I don't understand this change. Doesn't this give you two copies of the `root_block`?

Comment (hanabi1224) [reply tests/car_tests.rs:107]: Hmm I did not notice `rng.clone()` did the trick to make sure `root_block` is rerenerated, reverted

File: src/lib.rs [hunk 14]
```diff
@@ -24,7 +24,7 @@ mod chain_sync;
 mod cli;
 mod cli_shared;
 mod daemon;
-mod db;
+pub mod db;
```
Comment (lemmih) [inline src/lib.rs:27]: `db` and `utils` should not be public. You most likely want to convert the integration test into a unit test.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 14 above)
Comment (hanabi1224) [reply src/lib.rs]: Moved.  Btw, since we are hosting forest CAR files now, what's the plan of providing libraries for manipulating forest CARs?

Comment (lemmih) [reply src/lib.rs]: There are third-party crates for reading and writing CAR files. I don't think we need to provide anything. The ForestCAR.zst format is internal, and I don't want other people to depend on it (for now).

**Review by lemmih — APPROVED** [review]

File: src/cli/subcommands/car_cmd.rs [hunk 15]
```diff
@@ -204,21 +193,144 @@ mod tests {
...
+        assert!(!reader.header.roots.is_empty());
+        let mut count = 0;
+        let blocks: Vec<_> = reader.inspect_ok(|_| count += 1).try_collect().await?;
+        assert!(blocks.iter().map(|b| b.cid).all_unique());
+        println!("Result car block count: {count}");
+        Ok(())
+    }
+
+    fn cli() -> Result<Command> {
+        Ok(Command::cargo_bin("forest-cli")?)
```
Comment (lemmih) [inline src/cli/subcommands/car_cmd.rs:334]: `cargo_bin` only works from the context of integration tests.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: src/cli/subcommands/car_cmd.rs (hunk 15 above)
Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs:334]: Does it? I can see these tests passing locally and on CI

Comment (lemmih) [reply src/cli/subcommands/car_cmd.rs:334]: It says so in the documentation: https://docs.rs/assert_cmd/latest/assert_cmd/cargo/index.html

Could you verify that it works even if the binaries aren't built first?

Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs:334]: `cargo test` builds binaries by default but if `--lib` flag is used it won't build binaries. We have a few options
1. Delete these tests.
2. Move back to tests/ and make `util` and `db` mod public
3. build `forest-cli` with `escargot` (but it takes more than 30s and will timeout)

Comment (hanabi1224) [reply src/cli/subcommands/car_cmd.rs]: Removed

**Comment by hanabi1224** [comment]
@lemmih Is there anything else outstanding to fix?

**Review by lemmih — APPROVED** [review]

**Pull request merged**